File: 2014-07-18-FP/SherlockAndTheMaze/satm.accepted.scala
object Solution {
    val mm: Long = 1000000007L

    sealed class F(n: Int, m: Int, k: Int) {
        lazy val f = ff(n, m, k)
    }

    val dpa =
        (for {
            n <- 1 to 100
            m <- 1 to 100
            k <- 0 to 100
        } yield (new F(n, m, k))).toArray

    def dp(ix: (Int, Int, Int)): F =
        dpa(ix._3 + (ix._2 - 1) * 101 + (ix._1 - 1) * 10100)

    def g(n: Int, m: Int, k: Int): Long = {
        if (n > 0 && m > 0 && k >= 0)
            dp((n, m, k)).f
        else
            0
    }

    def ff(n: Int, m: Int, k: Int): Long = {
        (n, m, k) match {
            case (_, 1, 0) => 1L
            case (1, 1, _) => 0L
            case (_, 1, _) => 0L
            case (1, _, _) => 0L
            case (n, m, k) => (g(n - 1, m, k) + g(m, n - 1, k - 1)) % mm
        }
    }

    def main(args: Array[String]) = {
        val t = readLine.toInt
        for (ix <- 1 to t) {
            val (n :: m :: k :: _) = readLine.split(" ").map(_.toInt).toList
            if (n == 1 && m == 1)
                println("1")
            else {
                val ss = for (jx <- 0 to k) yield (dp((n, m, jx)).f + dp((m, n, jx)).f)
                println((ss foldLeft 0L)((a, b) => (a + b) % mm))
            }
        }
    }
}

File: 2014-07-18-FP/SherlockAndTheMaze/satm.accepted.fs
open System

let mm = 1000000007L

let rec dpa: Lazy<int64> array = [| for n in 1 .. 100 do for m in 1 .. 100 do for k in 0 .. 100 -> lazy (f (n, m, k)) |]
and f =
    function
    | (_, 1, 0) -> 1L
    | (_, 1, _) | (1, _, _) -> 0L
    | (n, m, k) -> ((g (n - 1) m k) + (g m (n - 1) (k - 1))) % mm
and dp n m k = dpa.[k + (m - 1) * 101 + (n - 1) * 10100]
and g n m k =
    if n > 0 && m > 0 && k >= 0
        then (dp n m k).Force ()
        else 0L

let s n m k = List.fold (fun a b -> (a + b) % mm) 0L (List.map (fun x -> (g n m x + g m n x) % mm) [0 .. k])

let tst _ =
    let (n :: m :: k :: _) = List.map int (List.ofSeq ((Console.ReadLine ()).Split [|' '|]))
    if n = 1 && m = 1
        then ignore (Console.WriteLine ("{0}", 1))
        else ignore (Console.WriteLine ("{0}", s n m k))

let main _ =
    let t = int (Console.ReadLine ())
    ignore (List.map tst [1 .. t])

main ()

